docs/architecture/psa-migration/psa-legacy-bridges.md

```diff
@@ -0,0 +1,321 @@
+* Persistence and location. This is completely orthogonal to the information from the `mbedtls_pk_context` object. It is convenient, but not necessary, for the conversion function to allow customizing these aspects. If it doesn't, users can call the conversion function and then call `psa_copy_key` and `psa_destroy_key` to move the key to its desired location.
+
+To allow the full flexibility around policies, and make the creation of a persistent key more convenient, the conversion function shall take a `const psa_key_attributes_t *` input, like all other functions that create a PSA key. In addition, there shall be a helper function to populate a `psa_key_attributes_t` with a sensible default. This lets the caller choose a more flexible, or just different usage policy, unlike the default-then-copy approach which only allows restricting the policy.
+
+This is close to the existing function `mbedtls_pk_wrap_as_opaque`, but does not bake in the implementation-specific consideration that a PSA key has exactly two algorithms, and also allows the caller to benefit from default for the policy in more cases.
+
+[ACTION] Implement `mbedtls_pk_get_psa_attributes` and `mbedtls_pk_import_into_psa` as described below. These functions are available whenever `MBEDTLS_PK_C` and `MBEDTLS_PSA_CRYPTO_CLIENT` are both defined. Deprecate `mbedtls_pk_wrap_as_opaque`.
+
+```
+int mbedtls_pk_get_psa_attributes(const mbedtls_pk_context *pk,
+                                  psa_key_attributes_t *attributes);
+int mbedtls_pk_import_into_psa(const mbedtls_pk_context *pk,
+                               const psa_key_attributes_t *attributes,
+                               mbedtls_svc_key_id_t *key_id);
+```
+
+* `mbedtls_pk_get_psa_attributes` does not change the id/lifetime fields of the attributes (which indicate a volatile key by default).
+* `mbedtls_pk_get_psa_attributes` sets the type and size based on what's in the pk context.
+    * The key type is a key pair if the context contains a private key, and a public key if the context only contains a public key.
+* `mbedtls_pk_get_psa_attributes` sets all the potentially applicable usage flags: `EXPORT`, `COPY`; `VERIFY_HASH | VERIFY_MESSAGE` or `ENCRYPT` as applicable for both public keys and key pairs; `SIGN` or `DECRYPT` as applicable for a key pair.
+* [OPEN] What is the default algorithm for `mbedtls_pk_get_psa_attributes`? Suggestion: assume signature by default. For RSA, either `PSA_RSA_PKCS1_V15_SIGN(PSA_ALG_ANY_HASH)` or `PSA_ALG_RSA_PSS(hash_alg)` depending on the RSA context's padding mode. For ECC, `PSA_ALG_DETERMINISTIC_ECDSA` if `MBEDTLS_ECDSA_DETERMINISTIC` is enabled and `PSA_ALG_ECDSA` otherwise.
+* [OPEN] Or does `mbedtls_pk_get_psa_attributes` need an extra argument indicating how to treat RSA and ECC keys?
```

Comment:
Since there are several reasonable ways to treat RSA keys and several reasonable ways to treat ECC keys, maybe this function needs an extra argument or two that specifies which of the reasonable treatments should be applied if the key is RSA, and which of the reasonable treatments should be applied if the key is ECC. For example: `MBEDTLS_PK_CONV_RSA_IS_PSS | MBEDTLS_PK_CONV_ECDSA_IS_DETERMINISTIC` means “if it's an RSA key then set its policy to `PSA_ALG_PSS(PSA_ALG_ANY_HASH)`, and if it's an ECDSA key then set its policy to `PSA_ALG_DETERMINISTIC_ECDSA(PSA_ALG_ANY_HASH)`”.

Reply:
Ah, I see, thanks for clarifying. No clear opinion right now, I'll think about it.

Reply:
Actually, do we really want `pk_get_psa_attributes()` (or however it's called) to do everything, or just what it actually can? How about it just sets key type and size, and we let users do the rest (alg + policy) using standard PSA APIs?

Otherwise we end up either:
- Using defaults / heuristics which will always be questionable, and when the result is not what the user wants they'll have to call `psa_set_key_algorithm()` and/or `psa_set_key_usage_flags()` anyway.
- Or asking the user and then having to develop a new API for that (flags like `MBEDTLS_PK_CONV_RSA_IS_PSS`) which is likely not to be as good as the existing PSA API.

Right now I'm inclined to keep it simple. So, instead of `pk_guess/suggest_psa_attributes()` we'd have `pk_get_psa_key_type_and_size()` + standard PSA APIs. IMO the fact this makes the the function is easier to name is a good sign. It could either still operate on an attributes structure, or just return type and size as separate output args, which might be easier for users who will likely want to make decisions based on the type then (or we could have a mix).

Wdyt?

Reply:
I do think it's part of the job of this function to construct a useful PSA key. That means it needs to set usage flags and at least a primary algorithm. Most simple applications should be able to call `mbedtls_pk_get_psa_attributes` followed by `mbedtls_pk_import_into_psa` without having to know how PSA policies work.

Reply:
> without having to know how PSA policies work.

Is that really a worthy goal? People are going to have to learn about PSA policies at some point anyway.

Reply:
It's not a goal, but if there's a common case where it's easy to achieve, that makes users' life easier.

Reply:
Makes sense. In that case I'm more inclined towards picking reasonable defaults (as suggested in the previous point) and then people who want something else can just call the standard PSA APIs to specify what they want.

Reply:
In https://github.com/Mbed-TLS/mbedtls/pull/8682 I tried to write this function with reasonable defaults. The result is somewhat questionable.